apps/election-manager/src/lib/votecounting.ts

```diff
@@ -487,70 +538,70 @@ export function filterTalliesByParams(
-  if (scannerId && precinctId) {
-    const precinctTally =
-      resultsByCategory.get(TallyCategory.Precinct)?.[precinctId] ||
-      getEmptyTally(precinctId)
-    const scannerTally =
-      resultsByCategory.get(TallyCategory.Scanner)?.[scannerId] ||
-      getEmptyTally(scannerId)
-    const cvrMap = new Map<string, CastVoteRecord>()
-    const allVotes: VotesDict[] = []
-    for (const ballotId of precinctTally.castVoteRecords.keys()) {
-      if (scannerTally.castVoteRecords.has(ballotId)) {
-        const CVR = precinctTally.castVoteRecords.get(ballotId)!
-        const vote = buildVoteFromCvr({ election, cvr: CVR })
-        cvrMap.set(ballotId, CVR)
-        allVotes.push(vote)
-      }
-    }
 
-    const contestTallies = tallyVotesByContest({ election, votes: allVotes })
-    const contestTallyMetadata = getContestTallyMeta({
-      election,
-      castVoteRecords: Object.values(cvrMap),
-    })
-    tallyResults = {
-      contestTallies,
-      castVoteRecords: cvrMap,
-      numberOfBallotsCounted: allVotes.length,
-      contestTallyMetadata,
-    }
-  } else if (scannerId) {
-    tallyResults =
+  if (!scannerId && !precinctId && !partyId) {
+    return overallTally
+  }
+
+  if (scannerId && !precinctId && !partyId) {
+    return (
       resultsByCategory.get(TallyCategory.Scanner)?.[scannerId] ||
-      getEmptyTally(scannerId)
-  } else if (precinctId) {
-    tallyResults =
+      getEmptyTally()
+    )
+  }
+
+  if (precinctId && !scannerId && !partyId) {
+    return (
       resultsByCategory.get(TallyCategory.Precinct)?.[precinctId] ||
-      getEmptyTally(precinctId)
+      getEmptyTally()
+    )
   }
-  if (!party) {
-    return tallyResults
+  if (partyId && !scannerId && !precinctId) {
+    return (
+      resultsByCategory.get(TallyCategory.Party)?.[partyId] || getEmptyTally()
+    )
   }
+  const cvrFiles: CastVoteRecord[] = []
+  const allVotes: VotesDict[] = []
 
-  // Filter By Party
-  const districts = election.ballotStyles
-    .filter((bs) => bs.partyId === party.id)
-    .flatMap((bs) => bs.districts)
-  const contestIds = expandEitherNeitherContests(
-    election.contests.filter(
-      (contest) =>
-        districts.includes(contest.districtId) && contest.partyId === party.id
-    )
-  ).map((contest) => contest.id)
+  const precinctTally =
+    precinctId && resultsByCategory.get(TallyCategory.Precinct)?.[precinctId]
+  const scannerTally =
+    scannerId && resultsByCategory.get(TallyCategory.Scanner)?.[scannerId]
+  const partyTally =
+    partyId && resultsByCategory.get(TallyCategory.Party)?.[partyId]
+
+  // TODO(#2975): Once we're removing duplicate ballots, make finding the intersection of these lists more performant.
+  overallTally.castVoteRecords.forEach((CVR) => {
+    if (!precinctTally || precinctTally.castVoteRecords.includes(CVR)) {
+      if (!scannerTally || scannerTally.castVoteRecords.includes(CVR)) {
+        if (!partyTally || partyTally.castVoteRecords.includes(CVR)) {
```

Comment:
If you constrain the type of the `*Tally` vars to `Tally | undefined` instead of `Tally | "" | undefined` you can use optional chaining. Use if you like:

```suggestion
  const precinctTally = precinctId
    ? resultsByCategory.get(TallyCategory.Precinct)?.[precinctId]
    : undefined
  const scannerTally = scannerId
    ? resultsByCategory.get(TallyCategory.Scanner)?.[scannerId]
    : undefined
  const partyTally = partyId
    ? resultsByCategory.get(TallyCategory.Party)?.[partyId]
    : undefined

  // TODO(#2975): Once we're removing duplicate ballots, make finding the intersection of these lists more performant.
  overallTally.castVoteRecords.forEach((CVR) => {
    if (precinctTally?.castVoteRecords.includes(CVR)) {
      if (scannerTally?.castVoteRecords.includes(CVR)) {
        if (partyTally?.castVoteRecords.includes(CVR)) {
```